src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,358 @@
+
+    public ValueTask<RegisterExtensionResponse> RegisterExtensionAsync(
+        Solution solution,
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        var analyzerAssemblyLoaderProvider = solution.Services.GetRequiredService<IAnalyzerAssemblyLoaderProvider>();
+
+        Extension? extension;
+        lock (_lockObject)
+        {
+            // Check if the assembly is already loaded.
+            if (!_extensions.TryGetValue(assemblyFolderPath, out extension))
+            {
+                var analyzerAssemblyLoader = analyzerAssemblyLoaderProvider.CreateNewShadowCopyLoader();
+
+                // Allow this assembly loader to load any dll in assemblyFolderPath.
+                foreach (var dll in Directory.EnumerateFiles(assemblyFolderPath, "*.dll"))
+                {
+                    try
+                    {
+                        // Check if the file is a valid .NET assembly.
+                        AssemblyName.GetAssemblyName(dll);
+                    }
+                    catch
+                    {
+                        // The file is not a valid .NET assembly, skip it.
+                        continue;
+                    }
+
+                    analyzerAssemblyLoader.AddDependencyLocation(dll);
+                }
+
+                extension = new Extension(this, analyzerAssemblyLoader, assemblyFolderPath);
+                _extensions[assemblyFolderPath] = extension;
+            }
+
+            if (extension.TryGetAssemblyHandlers(assemblyFileName, out var assemblyHandlers))
+            {
+                if (assemblyHandlers is null)
+                {
+                    throw new InvalidOperationException($"A previous attempt to load {assemblyFilePath} failed.");
+                }
+
+                return ValueTask.FromResult<RegisterExtensionResponse>(new(
+                    assemblyHandlers.WorkspaceMessageHandlers.Keys.ToImmutableArray(),
+                    assemblyHandlers.DocumentMessageHandlers.Keys.ToImmutableArray()));
+            }
+        }
+
+        return extension.LoadAssemblyAsync(assemblyFileName);
+    }
+
+    public ValueTask<string> HandleExtensionWorkspaceMessageAsync(Solution solution, string messageName, string jsonMessage, CancellationToken cancellationToken)
+        => HandleExtensionMessageAsync(solution, messageName, jsonMessage, _workspaceHandlers, cancellationToken);
+
+    public ValueTask<string> HandleExtensionDocumentMessageAsync(Document document, string messageName, string jsonMessage, CancellationToken cancellationToken)
+        => HandleExtensionMessageAsync(document, messageName, jsonMessage, _documentHandlers, cancellationToken);
```

Comment:
please place HandleExtensionMessageAsync here.  or move these lower down next to that helper.

Reply:
Moving them down just before the private `HandleExtensionMessageAsync `. Just trying to keep public methods before private ones due to PTSD (in other repos the analyzers complain if that is not the case)